src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+                deactivation.success(())
+              }
+
+              override def userEventTriggered(ctx: ChannelHandlerContext, evt: Any): Unit = {
+                evt match {
+                  case e: SslHandshakeCompletionEvent =>
+                    if (e.isSuccess) activation.success(ctx) else activation.failure(e.cause())
+                    log.debug(
+                      s"Ssl Handshake client channel from ${ctx.channel().localAddress()} " +
+                        s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id} and ssl status ${e.isSuccess}"
+                    )
+                  case _ =>
+                    log.debug(
+                      s"User Event client channel from ${ctx.channel().localAddress()} " +
+                        s"to ${ctx.channel().remoteAddress()} with channel id ${ctx.channel().id}"
+                    )
+                }
+                super.userEventTriggered(ctx, evt)
+              }
+            })
+            .addLast(new MessageNotifier[M](messageSubject))
+        }
+      })
+
+    def initialize: Task[ClientChannelImpl[M]] = {
+      toTask(bootstrap.connect(inetSocketAddress)).map(_ => this)
+    }
+
+    override def sendMessage(message: M): Task[Unit] = {
+
+      Task
+        .fromFuture(activationF)
+        .map(ctx => {
+          log.debug(
+            s"Processing outbound message from local address ${ctx.channel().localAddress()} " +
+              s"to remote address ${ctx.channel().remoteAddress()} via channel id ${ctx.channel().id()}"
+          )
+
+          ctx.writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
+        })
+        .map(_ => ())
+    }
+
+    override def in: Observable[M] = messageSubject
+
+    override def close(): Task[Unit] = {
+      messageSubject.onComplete()
+      Task
+        .fromFuture(activationF)
+        .flatMap(ctx => toTask(ctx.close()))
+        .flatMap(_ => Task.fromFuture(deactivationF))
+    }
+  }
+
+  private val channelSubject = PublishSubject[Channel[InetMultiAddress, M]]()
+
+  private val workerGroup = new NioEventLoopGroup()
+
+  private val clientBootstrap = new Bootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioSocketChannel])
+    .option[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
+    .option[RecvByteBufAllocator](ChannelOption.RCVBUF_ALLOCATOR, new DefaultMaxBytesRecvByteBufAllocator)
+    .option[Integer](ChannelOption.CONNECT_TIMEOUT_MILLIS, 10000)
+
+  private val serverBootstrap = new ServerBootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioServerSocketChannel])
+    .childHandler(new ChannelInitializer[SocketChannel]() {
+      override def initChannel(ch: SocketChannel): Unit = {
+        val newChannel = new ServerChannelImpl[M](ch)
+        channelSubject.onNext(newChannel)
+        log.debug(s"$processAddress received inbound from ${ch.remoteAddress()}.")
```

Comment:
It can be a good idea to log before feeding the subject, one advantage would be to simplify tracing, receiving the item and seeing the log after that could cause confusion.